sw/device/lib/base/bitfield.h

```diff
@@ -13,58 +14,101 @@ extern "C" {
  *
- * This function uses the `field` parameter to get a value at the given
- * offset (and mask) in `bitfield`. Only `mask` and `index` are used;
- * `value` is ignored.
+ * This function uses the `field` parameter to read the value from `bitfield`.
+ * The resulting value will be unshifted and zero-extended so the field's
+ * zero-bit is in the return value's zero-bit.
  *
  * @param bitfield Bitfield to get the field from.
  * @param field Field to read out from.
- * @return The read value, zero-extended to 32 bits.
+ * @return Zero-extended `field` from `bitfield`.
  */
-inline uint32_t bitfield_get_field32(uint32_t bitfield,
-                                     bitfield_field32_t field) {
+inline uint32_t bitfield_field32_read(uint32_t bitfield,
+                                      bitfield_field32_t field) {
   return (bitfield >> field.index) & field.mask;
 }
 
 /**
- * Sets `field` in `bitfield`.
+ * Writes `value` to `field` in `bitfield`.
  *
- * This function uses the `field` parameter to set a value
- * at a given offset in `bitfield`. The relevant portion of `bitfield`
- * is zeroed before the value is set.
+ * This function uses the `field` parameter to set specific bits in `bitfield`.
+ * The relevant portion of `bitfield` is zeroed before the bits are set to
+ * `value`.
  *
  * @param bitfield Bitfield to set the field in.
- * @param field Field within selected register field to be set.
- * @return The 32-bit resulting bitfield.
+ * @param field Field within bitfield to be set.
+ * @param value Value for the new field.
+ * @return `bitfield` with `field` set to `value`.
  */
-inline uint32_t bitfield_set_field32(uint32_t bitfield,
-                                     bitfield_field32_t field) {
+inline uint32_t bitfield_field32_write(uint32_t bitfield,
+                                       bitfield_field32_t field,
+                                       uint32_t value) {
   bitfield &= ~(field.mask << field.index);
-  bitfield |= (field.value & field.mask) << field.index;
+  bitfield |= (value & field.mask) << field.index;
+  return bitfield;
+}
+
+/**
+ * A single bit in a 32-bit bitfield.
+ *
+ * This denotes the position of a single bit, counting from the zero-bit.
+ *
+ * For instance, `(bitfield_bit_index_t)4` denotes the X-marked bit in the
+ * following 32-bit bitfield:
+ *
+ *     field:  0b--------'--------'--------'---X----
+ *     index:   31                                 0
+ *
+ * Restrictions: The value must not be greater than 31.
+ */
+typedef uint32_t bitfield_bit32_index_t;
+
+/**
+ * Reads the `bit_index`th bit in `bitfield`.
+ *
+ * @param bitfield Bitfield to get the bit from.
+ * @param bit_index Bit to read.
+ * @return `true` if the bit was one, `false` otherwise.
+ */
+inline bool bitfield_bit32_read(uint32_t bitfield,
+                                bitfield_bit32_index_t bit_index) {
+  return (bitfield & (0x1u << bit_index)) == (0x1u << bit_index);
```

Comment:
I think `(bitfield >> bit_index) & 0x1)` this is closer to what people normally do, and hence makes it clearer for the reader. In the end it doesn't matter, it is a pretty simple function.

It's just if there is a practical justification of doing then sure, otherwise it just adding more complexity (admittedly not much), for no reason.

Reply:
What about `(bitfield & (0x1u << bit_index)) != 0` to avoid duplicating the shift? This is also more clear imo.

Reply:
I actually agree with alphan. I expected this to be implemented in terms of field32.

Reply:
> Can't we just call bitfield_field32_read instead?

You mean internally in the `bitfield_bit32_read`?

Reply:
> > Can't we just call bitfield_field32_read instead?
> 
> You mean internally in the `bitfield_bit32_read`?

Yes, my comment is for this line (line 96 in this commit).

Reply:
Refactored to use `bitfield_field32_read`. Update incoming.